# First, let's inspect and clean our CitiBike trips data

In [7]:
import pandas as pd

df_list = []

# Getting DataFrames from our .csv files using loop
for i in range(1,13):
    if i < 10:
        df_list.append(pd.read_csv('data/JC-20230' + str(i) + '-citibike-tripdata.csv'))
    else:
        df_list.append(pd.read_csv('data/JC-2023' + str(i) + '-citibike-tripdata.csv'))
combined_df = pd.concat(df_list)
combined_df.reset_index(drop=True, inplace=True)
# Let's see how many rows we got
print('Total number of rows in Citi Bike dataset: ' + str(combined_df.shape[0]))
combined_df.head(988851)

Total number of rows in Citi Bike dataset: 988851


ride_id  rideable_type           started_at  \
0       0905B18B365C9D20   classic_bike  2023-01-28 09:18:10   
1       B4F0562B05CB5404  electric_bike  2023-01-23 20:10:12   
2       5ABF032895F5D87E   classic_bike  2023-01-29 15:27:04   
3       E7E1F9C53976D2F9   classic_bike  2023-01-24 18:35:08   
4       323165780CA0734B   classic_bike  2023-01-21 20:44:09   
...                  ...            ...                  ...   
988846  7E862701EE7A6A03   classic_bike  2023-12-20 17:55:41   
988847  DFAF91AB91BE25DB   classic_bike  2023-12-13 14:36:43   
988848  AEE7B49E6EDFAE6F   classic_bike  2023-12-18 16:40:00   
988849  A38E98956AD72EFB   classic_bike  2023-12-02 13:42:51   
988850  F16E0B969D877368   classic_bike  2023-12-19 16:30:13   

                   ended_at                        start_station_name  \
0       2023-01-28 09:28:52  Hoboken Terminal - Hudson St & Hudson Pl   
1       2023-01-23 20:18:27  Hoboken Terminal - Hudson St & Hudson Pl   
2       2023-01-29 15:32:38  Hoboken Terminal - Hudson St & Hudson Pl   
3       2023-01-24 18:42:13  Hoboken Terminal - Hudson St & Hudson Pl   
4       2023-01-21 20:48:08                             Hamilton Park   
...                     ...                                       ...   
988846  2023-12-20 18:00:10                          Heights Elevator   
988847  2023-12-13 14:42:31                          Heights Elevator   
988848  2023-12-18 16:45:45                           6 St & Grand St   
988849  2023-12-02 13:50:42                           6 St & Grand St   
988850  2023-12-19 16:36:54                           6 St & Grand St   

       start_station_id                            end_station_name  \
0                 HB101                               Hamilton Park   
1                 HB101  Southwest Park - Jackson St & Observer Hwy   
2                 HB101                          Marshall St & 2 St   
3                 HB101                               Hamilton Park   
4                 JC009                                Manila & 1st   
...                 ...                                         ...   
988846            JC059                       Bloomfield St & 15 St   
988847            JC059                       Bloomfield St & 15 St   
988848            HB302                       Bloomfield St & 15 St   
988849            HB302                       Bloomfield St & 15 St   
988850            HB302                           Clinton St & 7 St   

       end_station_id  start_lat  start_lng    end_lat    end_lng  \
0               JC009  40.735938 -74.030305  40.727596 -74.044247   
1               HB401  40.735938 -74.030305  40.737551 -74.041664   
2               HB408  40.735944 -74.030383  40.740802 -74.042521   
3               JC009  40.735986 -74.030364  40.727596 -74.044247   
4               JC082  40.727596 -74.044247  40.721651 -74.042884   
...               ...        ...        ...        ...        ...   
988846          HB203  40.748720 -74.040487  40.754530 -74.026580   
988847          HB203  40.748767 -74.040470  40.754530 -74.026580   
988848          HB203  40.744291 -74.034404  40.754530 -74.026580   
988849          HB203  40.744398 -74.034501  40.754530 -74.026580   
988850          HB303  40.744398 -74.034501  40.745420 -74.033320   

       member_casual  
0             member  
1             member  
2             member  
3             member  
4             member  
...              ...  
988846        casual  
988847        member  
988848        member  
988849        member  
988850        member  

[988851 rows x 13 columns]

# Examining null values

In [8]:
combined_df.isnull().sum()


ride_id                  0
rideable_type            0
started_at               0
ended_at                 0
start_station_name      85
start_station_id        85
end_station_name      3319
end_station_id        3319
start_lat                0
start_lng                0
end_lat                954
end_lng                954
member_casual            0
dtype: int64

# As we can see there are missing 3319 values in columns end_station_name and end_station_id.
# Let's figure out what can be the reason of that and how we can tackle this issue.

In [9]:
combined_df[combined_df['end_station_name'].isnull()] 

ride_id  rideable_type           started_at  \
448     8EA045BBA37A0719   classic_bike  2023-01-06 17:51:41   
547     035683025D51ACA0   classic_bike  2023-01-28 23:03:42   
549     ECA9E0C8111549F0  electric_bike  2023-01-21 01:45:25   
567     68759C3009682643   classic_bike  2023-01-27 09:01:26   
569     A7441B9B51909946   classic_bike  2023-01-13 16:03:30   
...                  ...            ...                  ...   
983559  8C1705634913B6E7   classic_bike  2023-12-18 14:12:25   
985729  EF861287AFC29122   classic_bike  2023-12-13 07:42:49   
985730  5C46CC1B233A9EE8   classic_bike  2023-12-08 08:13:25   
987479  F9508E6CFB511D70   classic_bike  2023-12-10 12:18:10   
987741  B9D3FDD2931A78D3   classic_bike  2023-12-27 05:21:37   

                   ended_at                        start_station_name  \
448     2023-01-06 17:51:45  Hoboken Terminal - Hudson St & Hudson Pl   
547     2023-01-30 00:03:34                             Hamilton Park   
549     2023-01-22 02:45:21                         Madison St & 1 St   
567     2023-01-27 09:06:37                             Hamilton Park   
569     2023-01-13 17:08:06                            Pershing Field   
...                     ...                                       ...   
983559  2023-12-19 15:12:20                Hoboken Ave at Monmouth St   
985729  2023-12-13 08:47:42                Hoboken Ave at Monmouth St   
985730  2023-12-08 09:30:59                                  Union St   
987479  2023-12-10 13:36:55                         Madison St & 1 St   
987741  2023-12-27 06:19:47                         Madison St & 1 St   

       start_station_id end_station_name end_station_id  start_lat  start_lng  \
448               HB101              NaN            NaN  40.736001 -74.030338   
547               JC009              NaN            NaN  40.727596 -74.044247   
549               HB402              NaN            NaN  40.738790 -74.039300   
567               JC009              NaN            NaN  40.727553 -74.044259   
569               JC024              NaN            NaN  40.742651 -74.051815   
...                 ...              ...            ...        ...        ...   
983559            JC105              NaN            NaN  40.735208 -74.046964   
985729            JC105              NaN            NaN  40.735286 -74.046814   
985730            JC051              NaN            NaN  40.718015 -74.083314   
987479            HB402              NaN            NaN  40.738697 -74.039263   
987741            HB402              NaN            NaN  40.738822 -74.039195   

        end_lat  end_lng member_casual  
448       40.74   -74.03        member  
547         NaN      NaN        member  
549         NaN      NaN        member  
567       40.73   -74.04        member  
569       40.73   -74.05        casual  
...         ...      ...           ...  
983559      NaN      NaN        casual  
985729    40.73   -74.03        member  
985730    40.73   -74.03        member  
987479    40.75   -74.03        casual  
987741    40.74   -74.03        member  

[3319 rows x 13 columns]

# As we can the data is Missing Completely At Random, so in this case it's better to just remove those rows because some of them show abnormal high trip duration.

In [10]:
combined_df.dropna(inplace=True)

# Creating "Duration" column.

In [11]:
# First, ensure that start_end and start_at are in datetime format
combined_df['started_at'] = pd.to_datetime(combined_df['started_at'])
combined_df['ended_at'] = pd.to_datetime(combined_df['ended_at'])

# Subtract ended_at from started_at to create the duration column
combined_df['duration_min'] = ((combined_df['ended_at'] - combined_df['started_at']).dt.total_seconds() / 60).round(2)
combined_df.to_csv('data/ridership.csv', index=False)
combined_df

ride_id  rideable_type          started_at  \
0       0905B18B365C9D20   classic_bike 2023-01-28 09:18:10   
1       B4F0562B05CB5404  electric_bike 2023-01-23 20:10:12   
2       5ABF032895F5D87E   classic_bike 2023-01-29 15:27:04   
3       E7E1F9C53976D2F9   classic_bike 2023-01-24 18:35:08   
4       323165780CA0734B   classic_bike 2023-01-21 20:44:09   
...                  ...            ...                 ...   
988846  7E862701EE7A6A03   classic_bike 2023-12-20 17:55:41   
988847  DFAF91AB91BE25DB   classic_bike 2023-12-13 14:36:43   
988848  AEE7B49E6EDFAE6F   classic_bike 2023-12-18 16:40:00   
988849  A38E98956AD72EFB   classic_bike 2023-12-02 13:42:51   
988850  F16E0B969D877368   classic_bike 2023-12-19 16:30:13   

                  ended_at                        start_station_name  \
0      2023-01-28 09:28:52  Hoboken Terminal - Hudson St & Hudson Pl   
1      2023-01-23 20:18:27  Hoboken Terminal - Hudson St & Hudson Pl   
2      2023-01-29 15:32:38  Hoboken Terminal - Hudson St & Hudson Pl   
3      2023-01-24 18:42:13  Hoboken Terminal - Hudson St & Hudson Pl   
4      2023-01-21 20:48:08                             Hamilton Park   
...                    ...                                       ...   
988846 2023-12-20 18:00:10                          Heights Elevator   
988847 2023-12-13 14:42:31                          Heights Elevator   
988848 2023-12-18 16:45:45                           6 St & Grand St   
988849 2023-12-02 13:50:42                           6 St & Grand St   
988850 2023-12-19 16:36:54                           6 St & Grand St   

       start_station_id                            end_station_name  \
0                 HB101                               Hamilton Park   
1                 HB101  Southwest Park - Jackson St & Observer Hwy   
2                 HB101                          Marshall St & 2 St   
3                 HB101                               Hamilton Park   
4                 JC009                                Manila & 1st   
...                 ...                                         ...   
988846            JC059                       Bloomfield St & 15 St   
988847            JC059                       Bloomfield St & 15 St   
988848            HB302                       Bloomfield St & 15 St   
988849            HB302                       Bloomfield St & 15 St   
988850            HB302                           Clinton St & 7 St   

       end_station_id  start_lat  start_lng    end_lat    end_lng  \
0               JC009  40.735938 -74.030305  40.727596 -74.044247   
1               HB401  40.735938 -74.030305  40.737551 -74.041664   
2               HB408  40.735944 -74.030383  40.740802 -74.042521   
3               JC009  40.735986 -74.030364  40.727596 -74.044247   
4               JC082  40.727596 -74.044247  40.721651 -74.042884   
...               ...        ...        ...        ...        ...   
988846          HB203  40.748720 -74.040487  40.754530 -74.026580   
988847          HB203  40.748767 -74.040470  40.754530 -74.026580   
988848          HB203  40.744291 -74.034404  40.754530 -74.026580   
988849          HB203  40.744398 -74.034501  40.754530 -74.026580   
988850          HB303  40.744398 -74.034501  40.745420 -74.033320   

       member_casual  duration_min  
0             member         10.70  
1             member          8.25  
2             member          5.57  
3             member          7.08  
4             member          3.98  
...              ...           ...  
988846        casual          4.48  
988847        member          5.80  
988848        member          5.75  
988849        member          7.85  
988850        member          6.68  

[985519 rows x 14 columns]

# Now we'll examine our weather data. We will try to find unnecessary columns and drop them. 

In [12]:
# Reading csv file into DataFrame
w_df = pd.read_csv('data/export.csv')
# Removing unnecessary columns
w_df.drop(['wpgt', 'tsun'], axis=1, inplace=True)
w_df.to_csv('data/weather.csv', index=False)
# No Rain: 0 mm
# Light Rain: Greater than 0 mm up to 2.5 mm per day
# Moderate Rain: Greater than 2.5 mm up to 10 mm per day
# Heavy Rain: Greater than 10 mm up to 50 mm per day
# Very Heavy Rain: Greater than 50 mm per day
# Extreme Rain: Significantly more than 50 mm in a day, often leading to flooding and other hydrological impacts.
w_df

date  tavg  tmin  tmax  prcp  snow   wdir  wspd    pres
0    2023-01-01  11.3   7.2  14.4   0.0   0.0  241.0  13.0  1012.6
1    2023-01-02   9.9   6.1  15.0   1.3   0.0  231.0   7.9  1019.5
2    2023-01-03  11.2   9.4  13.3  11.4   0.0    4.0   5.4  1014.1
3    2023-01-04  13.7  10.0  18.9   0.5   0.0  164.0   9.0  1010.0
4    2023-01-05  11.4   8.3  12.8   0.0   0.0   57.0   6.8  1013.5
..          ...   ...   ...   ...   ...   ...    ...   ...     ...
360  2023-12-27   8.1   7.2  10.6   8.6   0.0   23.0   9.4  1017.8
361  2023-12-28  10.7  10.0  12.8  38.4   0.0   20.0  16.9  1005.5
362  2023-12-29  10.7   7.8  13.9   1.3   NaN  273.0  12.2  1004.5
363  2023-12-30   7.1   2.8   8.3   0.0   0.0  269.0  19.8  1004.6
364  2023-12-31   6.0   4.4   6.7   0.0   0.0  280.0  13.6  1014.6

[365 rows x 9 columns]

# Now our data is ready. There was no missing data that can mislead us so I left all rows untouched.